File: 35. Search Insert Position.go
package main

func binarySearch(nums []int, target, left, right int) int {
	if left <= right {

		mid := left + (right-left)/2

		if nums[mid] == target {
			return mid
		}

		if nums[mid] < target {
			return binarySearch(nums, target, mid+1, right)
		} else {
			return binarySearch(nums, target, left, mid-1)
		}
	}
	return left
}

func searchInsert(nums []int, target int) int {
	return binarySearch(nums, target, 0, len(nums)-1)
}

File: 35. Search Insert Position.swift
import Foundation

class Solution {
    
    var nums: Array<Int> = []
    
    private func binarySearch(_ left: Int,_ right: Int,_ target: Int) -> Int {
        
        if left <= right {
            let mid = left + (right - left)/2
            
            if nums[mid] == target {
                return mid
            }
            
            if nums[mid] > target {
                return binarySearch(left, mid-1, target)
            }else{
                return binarySearch(mid+1, right, target)
            }
        }
    
        return left
    }
    
    func searchInsert(_ nums: [Int], _ target: Int) -> Int {
        self.nums = nums
        return binarySearch(0, nums.count-1, target)
    }
}